### PR TITLE
Feat: Introduce state backups during migrations

### LINKED ISSUE #684: Automatic backup before migration
When running a migration on a non-transactional db, failures can be unrecoverable. Migrate should create a backup of all tables before a migration, which can then be easily roll backed to

migrate() -> (make backup, run migration)

rollback(backup) -> (reset the state sync)

### PR DESCRIPTION
**Pull request by vchan**
Closes #684

### CODE REVIEW

File: sqlmesh/core/state_sync/engine_adapter.py [hunk 1]
```diff
@@ -401,9 +411,98 @@ def _get_environment(
         env = self._environment_from_row(row)
         return env
 
+    def _backup_table(self, schema_version: int, table_name: str) -> None:
+        data = self.engine_adapter.fetchdf(exp.select("*").from_(table_name))
+        self.engine_adapter.insert_append(
+            self.backups_table,
+            next(
```
Comment (izeigerman) [inline sqlmesh/core/state_sync/engine_adapter.py:418]: Wouldn't it be a little too much storing entire tables as a single row? Should we consider just copying tables instead?

File: sqlmesh/core/state_sync/engine_adapter.py [hunk 2]
```diff
@@ -401,9 +411,98 @@ def _get_environment(
         env = self._environment_from_row(row)
         return env
 
+    def _backup_table(self, schema_version: int, table_name: str) -> None:
```
Comment (izeigerman) [inline sqlmesh/core/state_sync/engine_adapter.py:414]: Should we also take into account the sqlglot version?

File: sqlmesh/core/state_sync/engine_adapter.py (hunk 1 above)
Comment (tobymao) [reply sqlmesh/core/state_sync/engine_adapter.py:418]: i was suggesting storing one column per row, but we could copy the tables

File: sqlmesh/core/state_sync/engine_adapter.py [hunk 3]
```diff
@@ -401,9 +411,98 @@ def _get_environment(
...
+
+    def _restore_table(
+        self,
+        schema_version: int,
+        table_name: str,
+        columns_to_types: t.Optional[t.Dict[str, exp.DataType]] = None,
+    ) -> None:
+        result = self.engine_adapter.fetchone(
+            exp.select(exp.to_column("data"))
+            .from_(self.backups_table)
```
Comment (tobymao) [inline sqlmesh/core/state_sync/engine_adapter.py:441]: where(exp.column(...).eq(schema_version)

File: sqlmesh/core/state_sync/engine_adapter.py (hunk 1 above)
Comment (vchan) [reply sqlmesh/core/state_sync/engine_adapter.py:418]: I gave it some thought and I think copying the tables and replacing them on rollback has the advantage of us not having to worry about a migration that alters one or more of the state tables.

File: sqlmesh/core/state_sync/engine_adapter.py (hunk 2 above)
Comment (vchan) [reply sqlmesh/core/state_sync/engine_adapter.py]: I included the sqlglot version in the backup metadata table.

File: sqlmesh/cli/main.py [hunk 4]
```diff
@@ -371,6 +371,19 @@ def migrate(ctx: click.Context) -> None:
     ctx.obj.migrate()
 
 
+@cli.command("rollback")
+@click.option(
+    "--schema-version",
```
Comment (tobymao) [inline sqlmesh/cli/main.py:376]: maybe just version

File: sqlmesh/core/context.py [hunk 5]
```diff
@@ -842,6 +842,13 @@ def migrate(self) -> None:
         """
         self._new_state_sync().migrate()
 
+    def rollback(self, schema_version: int = 0) -> None:
```
Comment (tobymao) [inline sqlmesh/core/context.py:845]: maybe get rid of all the default values here

File: sqlmesh/core/state_sync/base.py [hunk 6]
```diff
@@ -371,21 +371,32 @@ def unpause_snapshots(
...
+
         for migration in migrations:
             logger.info(f"Applying migration {migration}")
             migration.migrate(self)
 
         self._migrate_rows()
         self._update_versions()
 
+    @abc.abstractmethod
+    def rollback(self, schema_version: int = 0) -> None:
```
Comment (tobymao) [inline sqlmesh/core/state_sync/base.py:393]: get rid of default value

File: sqlmesh/core/state_sync/engine_adapter.py [hunk 7]
```diff
@@ -401,9 +411,96 @@ def _get_environment(
...
+        schema_version: int,
+        table_name: str,
+        columns_to_types: t.Optional[t.Dict[str, exp.DataType]] = None,
+    ) -> None:
+        result = self.engine_adapter.fetchone(
+            exp.select("table_name", "backup_table_name")
+            .from_(self.backups_table)
+            .where(
+                exp.to_column("schema_version")
+                .eq(exp.Literal.number(schema_version))
```
Comment (tobymao) [inline sqlmesh/core/state_sync/engine_adapter.py:445]: can you just do .eq(schema_vzersion)

File: sqlmesh/core/state_sync/engine_adapter.py [hunk 8]
```diff
@@ -401,9 +411,96 @@ def _get_environment(
         env = self._environment_from_row(row)
         return env
 
+    def _backup_table(self, versions: Versions, table_name: str) -> None:
```
Comment (tobymao) [inline sqlmesh/core/state_sync/engine_adapter.py:414]: what is this table for?

File: sqlmesh/core/state_sync/engine_adapter.py [hunk 9]
```diff
@@ -401,9 +411,96 @@ def _get_environment(
...
-    def migrate(self) -> None:
-        super().migrate()
+    def rollback(self, schema_version: int = 0) -> None:
+        if not self.engine_adapter.table_exists(self.backups_table):
+            return
+
+        if schema_version == 0:
+            # Get most recently backed up schema version, i.e. the previous schema version
+            (schema_version,) = self.engine_adapter.fetchone(
+                exp.select(exp.Max(this=exp.to_column("schema_version"))).from_(self.backups_table)
```
Comment (tobymao) [inline sqlmesh/core/state_sync/engine_adapter.py:473]: exp.select("MAX(schema_version)")

File: sqlmesh/core/state_sync/engine_adapter.py [hunk 10]
```diff
@@ -401,9 +411,96 @@ def _get_environment(
...
+        self._restore_table(
+            schema_version, self.environments_table, self.environment_columns_to_types
+        )
+        self._restore_table(schema_version, self.versions_table, self.version_columns_to_types)
+
+        # Clean up orphaned backups
+        for (backup_table_name,) in self.engine_adapter.fetchall(
+            exp.select("backup_table_name")
+            .from_(self.backups_table)
+            .where(exp.to_column("schema_version") > exp.Literal.number(schema_version))
```
Comment (tobymao) [inline sqlmesh/core/state_sync/engine_adapter.py:488]: where("schema_version > {schema_version}") or
where(exp.column("schema_version") > schema_version)

**Comment by tobymao** [comment]
ok, last thing is make sure this works on various engines, spark, bigquery, snowflake

File: sqlmesh/core/context.py (hunk 5 above)
Comment (tobymao) [reply sqlmesh/core/context.py:845]: this is still here

Comment (vchan) [reply sqlmesh/core/context.py:845]: Thanks, I forgot to add the file in git before committing.

**Review by tobymao — APPROVED** [review]

**Review by tobymao — APPROVED** [review]

**Pull request merged**